File: path-sum-iii/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int target;
    int count = 0;
    HashMap<Long, Integer> map = new HashMap<>();
    public int pathSum(TreeNode root, int targetSum) {
        target = targetSum;
        dfs(root, 0);
        return count;
    }

    private void dfs(TreeNode node, long sumSoFar) {
        if (node == null) {
            return;
        }
        sumSoFar += node.val;
        if (sumSoFar == target) {
            count++;
        }
        count += map.getOrDefault(sumSoFar - target, 0);

        map.put(sumSoFar, map.getOrDefault(sumSoFar, 0) + 1);

        dfs(node.left, sumSoFar);
        dfs(node.right, sumSoFar);

        map.put(sumSoFar, map.get(sumSoFar) - 1);
    }
}

File: 437-path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        prefix_sum = {}
        prefix_sum[0] = 1
        count = 0
        def explore(root, sum_so_far):
            nonlocal count
            if not root: return 0
            sum_so_far += root.val
            target = sum_so_far - targetSum
            if target in prefix_sum.keys():
                count += prefix_sum[target]
            if sum_so_far in prefix_sum:
                prefix_sum[sum_so_far] += 1
            else:
                prefix_sum[sum_so_far] = 1
            explore(root.left, sum_so_far)
            explore(root.right, sum_so_far)
            prefix_sum[sum_so_far] -= 1
        explore(root, 0)
        return count